### PR TITLE
Make minmax faster for Float32/64

### PR DESCRIPTION
**Pull request by N5N3**
1. ~This PR tries to fix `minmax` for BigFloat, as:~ The allocations in `BigFloat`'s `min/max` seems disputable. This PR just makes the outputs of `minmax` consistent with that of `min/max`.
```julia
julia> a = BigFloat(1//3, 300); b = BigFloat(3//5, 300);
julia> min(a, b) === a
false    -->  false
julia> max(a, b) === b
false    -->  false
julia> minmax(a, b) === (a, b)
true     -->  false
julia> min(a, b) == a
false    -->  false
julia> max(a, b) == b
false    -->  false
julia> minmax(a, b) == (a, b)
true     -->  false
```

2. At present, if the two inputs are NaNs with different signs:
- Base.min always returns the one with positive sign
- Base.max always returns the one with negetive sign
- Base.minmax(x, y) always returns (x, x)
I just wonder which style is better? This PR make min max also return the first of two NaNs (no matter their signs) like LLVM and Julia's BigFloat.

3. This PR also makes `min/max` a little faster for `IEEEFloat`.
At the same time, `minmax` for `Float32/64` could be proper vectorlized with broadcast now.
Some benchmarks:
```julia
julia> a = randn(100); b = randn(100); c = similar(a); d = minmax.(a,b); @btime $c .= min.($a,$b); @btime $c .= max.($a,$b); @btime $d .= minmax.($a,$b); 
  43.448 ns (0 allocations: 0 bytes)   --->   33.494 ns (0 allocations: 0 bytes)
  44.040 ns (0 allocations: 0 bytes)   --->   37.803 ns (0 allocations: 0 bytes)
  233.696 ns (0 allocations: 0 bytes)  --->   52.584 ns (0 allocations: 0 bytes)

julia> a = randn(Float32, 100); b = randn(Float32, 100); c = similar(a); d = minmax.(a,b); @btime $c .= min.($a,$b); @btime $c .= max.($a,$b); @btime $d .= minmax.($a,$b);
  29.819 ns (0 allocations: 0 bytes)   --->    24.172 ns (0 allocations: 0 bytes)
  29.789 ns (0 allocations: 0 bytes)   --->    23.447 ns (0 allocations: 0 bytes)
  223.353 ns (0 allocations: 0 bytes)  --->    40.664 ns (0 allocations: 0 bytes)

julia> a = randn(Float16, 100); b = randn(Float16, 100); c = similar(a); d = minmax.(a,b); @btime $c .= min.($a,$b); @btime $c .= max.($a,$b); @btime $d .= minmax.($a,$b);
  239.862 ns (0 allocations: 0 bytes)  --->   135.865 ns (0 allocations: 0 bytes)
  243.970 ns (0 allocations: 0 bytes)  --->   136.424 ns (0 allocations: 0 bytes)
  206.833 ns (0 allocations: 0 bytes)  --->   200.500ns (0 allocations: 0 bytes)
```

### CODE REVIEW

**Comment by Seelengrab** [comment]
The speedup for `Flaot32/64` is nice, but I'm not sure why allocating more for `BigFloat` is more correct? I'd have expected a non-allocating version to be more in line with what `min` and `max` are supposed to do, namely returning the min/max of its arguments (not a copy thereof).

**Comment by N5N3** [comment]
Since `BigFloat` is introduced in `Base.MPFR`, maybe such change should be done in a seperated PR?
`Base.Math` only offers the generic version of `AbstractFloat`'s `min/max/minmax`, and make them fast for `IEEEFloat`.

**Comment by Seelengrab** [comment]
If the change in allocations for `BigFloat` wasn't intended and happened incidentally, I'd be in favor of just doing it in this PR since it touches the same (user facing) API, touching a different file seems fine with me. I think it'd just be a method specialization for `BigFloat` with a non-allocating version, no? Does `MPFR` expose some `min`/`max` API we can use directly without having to implement it ourselves?

**Comment by N5N3** [comment]
I'm not familiar with `libmpfr`, but Julia now use the 4 args api (2 inputs + output + rounding setting) to calculate the min/max of `BigFloat`, and the following code return false, as the default prec is 256.
```julia
julia> a = BigFloat(1//3,300); b = BigFloat(1//5,300);
julia> min(a, b) == b
false
```
It seems to be a better example to motivate us to the change?

**Comment by N5N3** [comment]
I think we'd better pick this PR up. @oscardssmith @tkf 
List for triage:
1. Do we really need allocation in `min`/`max` for `BigFloat`.
2. Should we treat `copysign(NaN, -1) < NaN` in `min/max` for `AbstractFloat`.

**Comment by tkf** [comment]
I consider `min`/`max` to be defined with respect to total orderings [^1] `isless` and `Base.isgreater` since this is a crucial property for parallelism. The result happens to coincide with IEEE 754 for floats and the "poisoning" semantics of `missing`. Since `min`/`max`/`minimum`/`maximum`/`extrema`/etc. provide the "fist match" semantics, I think it makes sense to reuse the input, even for BigFloat NaNs.

[^1]: Or maybe more precisely "partial orderings that are total on the many of the subsets of values people tend to care"

**Comment by N5N3** [comment]
`BigFloat` is a little bit complex because of it's "variable" precison.
Should we do promotion in `min(x::BigFloat, y::BigFloat)` if they have different precision?

**Comment by tkf** [comment]
Hmm.... can the promotion of the precision be added to `promote` on `BigFloat`s? Though I guess that's too breaking...

**Comment by N5N3** [comment]
If we do that, I guess all other math calls should be modified:
```julia
julia> b = BigFloat(1.0,3000) + BigFloat(1.0,200)
2.0
julia> b.prec
256
```
Maybe we can avoid allocation in `min`/`max` only when all inputs has the default precision?

**Comment by tkf** [comment]
Actually, I have no idea what users would expect about the output precision of `BigFloat` operations. I'm no expert so I'll shut up :)

**Comment by N5N3** [comment]
I personally seldom use `BigFloat`. So let's wait the triage result \_(:зゝ∠)_.

**Comment by JeffBezanson** [comment]
From triage: (1) We don't need to copy a BigFloat, ok to return the actual element we find. (2) It doesn't matter which NaN we return when there are multiple.

**Comment by oscardssmith** [comment]
For `BigFloat` there's no reason to promote (it would just be slower and not more useful). The last thing to check is that this does correct things with `-0.0`. Is this code path sufficiently tested?

**Comment by N5N3** [comment]
I extend the current `min/max/minmax` test for `Float64` to all float types in `Base`. (5dd01a23bada5afc5fd08afbfea6861e939c800b)
Local test shows no problem.

`_extrema_rf` related optimization is also added in this PR, since it's `min/max` related.

**Comment by N5N3** [comment]
Update the lastest Benchmark:
```julia
for T in (Float32, Float64, Float16, BigFloat)
    a = T.(randn(128)); b = T.(randn(128)); c = similar(a); 
    d = minmax.(a,b); 
    t1 = @benchmark $c .= min.($a,$b); 
    t2 = @benchmark $c .= max.($a,$b); 
    t3 = @benchmark $d .= minmax.($a,$b);
    print(T, "|  min: ", round(median(t1).time, digits = 2), "ns")
    print("  max: ", round(median(t2).time, digits = 2), "ns")
    println("  minmax: ", round(median(t3).time, digits = 2), "ns")
end
```
This PR:
```julia
Float32|  min: 21.34ns  max: 21.26ns  minmax: 40.73ns
Float64|  min: 34.07ns  max: 34.14ns  minmax: 69.43ns
Float16|  min: 173.6ns  max: 174.18ns  minmax: 223.15ns
BigFloat|  min: 1610.0ns  max: 1540.0ns  minmax: 1630.0ns
```
Master:
```julia
Float32|  min: 28.82ns  max: 28.92ns  minmax: 231.8ns
Float64|  min: 46.87ns  max: 46.77ns  minmax: 243.28ns
Float16|  min: 317.72ns  max: 318.14ns  minmax: 262.72ns
BigFloat|  min: 5216.67ns  max: 5200.0ns  minmax: 2588.89ns
```

**Comment by N5N3** [comment]
All test passed except for `analyzegc`. Should be unrelated.

File: base/math.jl [hunk 1]
```diff
@@ -758,17 +758,34 @@ end
...
-    ifelse(isnan(x) | isnan(y), ifelse(isnan(x), (x,x), (y,y)),
-           ifelse((y > x) | (signbit(x) > signbit(y)), (x,y), (y,x)))
+function max(x::T, y::T) where {T<:Union{Float32,Float64}}
+    diff = x - y
+    argmax = ifelse(signbit(diff), y, x)
+    anynan = isnan(x)|isnan(y)
+    ifelse(anynan, diff, argmax)
+end
 
+function minmax(x::T, y::T) where {T<:Union{Float32,Float64}}
```
[the remark this answers is not in the document]
Comment (N5N3) [reply base/math.jl:782]: My local bench does show some performance difference.
```julia
julia> a = randn(1024); b = randn(1024); z = min.(a,b); zz = minmax.(a,b);

julia> using BenchmarkTools
[ Info: Precompiling BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @benchmark $zz .= minmax.($a, $b)
BenchmarkTools.Trial: 10000 samples with 198 evaluations.
 Range (min … max):  440.404 ns …  1.189 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     442.424 ns              ┊ GC (median):    0.00%        
 Time  (mean ± σ):   457.627 ns ± 45.207 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █    
  █▂▃▁▅▃▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁
  440 ns          Histogram: frequency by time          702 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> f(x, y) = min(x, y), max(x, y)
f (generic function with 1 method)

julia> @benchmark $zz .= f.($a, $b)
BenchmarkTools.Trial: 10000 samples with 194 evaluations.
 Range (min … max):  497.423 ns …  3.276 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     498.969 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   515.506 ns ± 55.169 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▄▂ ▆▅▄▂▁                                                    ▁
  ██████████▇▆▆▅▄▃▆▆▆▅▄▅▄▄▅▃▅▅▃▃▃▂▄▃▃▃▄▃▄▄▄▄▄▄▄▄▄▄▄▅▄▅▄▅▄▅▄▄▃▄ █
  497 ns        Histogram: log(frequency) by time       759 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**Comment by oscardssmith** [comment]
How big is the regression in speed on M1 for this? If it's not major, I'd be in favor of merging.

File: base/math.jl [hunk 2]
```diff
@@ -758,17 +758,34 @@ end
...
-min(x::T, y::T) where {T<:AbstractFloat} = ifelse((y < x) | (signbit(y) > signbit(x)),
-                                    ifelse(isnan(x), x, y), ifelse(isnan(y), y, x))
+_isless(x::T, y::T) where {T<:AbstractFloat} = (x < y) || (signbit(x) > signbit(y))
+min(x::T, y::T) where {T<:AbstractFloat} = isnan(x) || ~isnan(y) && _isless(x, y) ? x : y
+max(x::T, y::T) where {T<:AbstractFloat} = isnan(x) || ~isnan(y) && _isless(y, x) ? x : y
+minmax(x::T, y::T) where {T<:AbstractFloat} = min(x, y), max(x, y)
+
+_isless(x::Float16, y::Float16) = signbit(widen(x) - widen(y))
+
+function min(x::T, y::T) where {T<:Union{Float32,Float64}}
```
Comment (mikmoore) [inline base/math.jl:768]: Should the new `min`, `max`, `minmax` signatures be expanded to include `Float16`?

Comment (JeffreySarnoff) [reply base/math.jl:768]: using `T<:IEEEFloat` everywhere that now uses 
`T<:Union{Float32,Float64}` is helpful.  Some systems have onchip support for Float16s, so this is a win there.  The systems that emulate Float16s support min, max, minmax, so the processing is at worst unchanged there.

**Comment by JeffreySarnoff** [comment]
Each of the functions `min(x,y)`, `max(x,y)`, `minmax(x,y)` use this:
   `anynan = isnan(x) | isnan(y)`

consider
```
for (T,U) in ((:Float64, :UInt64). (:Float32, :UInt32), (:Float16, :UInt16))
  @eval anynan(x::$T, y::$T) =
    isnan( reinterpret($T, 
      ( reinterpret($U, x) | reinterpret($U, y)) 
    )
end
```

**Comment by mikmoore** [comment]
> Each of the functions `min(x,y)`, `max(x,y)`, `minmax(x,y)` use this: `anynan = isnan(x) | isnan(y)`
> 
> consider
> 
> ```
> for (T,U) in ((:Float64, :UInt64). (:Float32, :UInt32), (:Float16, :UInt16))
>   @eval anynan(x::$T, y::$T) =
>     isnan( reinterpret($T, 
>       ( reinterpret($U, x) | reinterpret($U, y)) 
>     )
> end
> ```

This will erroneously report true for many input combinations such as `anynan(Inf,1.5)`. But that aside, the `isnan(x)|isnan(y)` evaluation can actually be evaluated in a single instruction on x86 (a check that `x` and `y` compare as unordered), so nothing else could be appreciably faster. I can't speak for other architectures.

**Comment by JeffreySarnoff** [comment]
Ok, there is a bitop fix for my error, but nevermind.
Does Julia or the LLVM use that single instruction version?

**Comment by mikmoore** [comment]
> Does Julia or the LLVM use that single instruction version?

The `code_native` for this `min` on `Float64`:
```
        vsubsd  %xmm1, %xmm0, %xmm2
        vmovq   %xmm2, %rax
        vmovapd %xmm1, %xmm3
        testq   %rax, %rax
        jns     .LBB0_2
# %bb.1:                                # %top
        vmovapd %xmm0, %xmm3
.LBB0_2:                                # %top
        vcmpordsd       %xmm1, %xmm0, %xmm0
        vblendvpd       %xmm0, %xmm3, %xmm2, %xmm0
        retq
```
The `vcmpordsd` and `vblendvpd` correspond to the `anynan` determination and `ifelse` selection, respectively.

**Comment by N5N3** [comment]
I'm planning to merge this at the weekend if CI passed and there's no other objection.

**Comment by JeffreySarnoff** [comment]
thank you for the work

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**